engines/native/system/system_test.go

```diff
@@ -17,21 +19,81 @@ import (
 	"github.com/taskcluster/taskcluster-worker/runtime/ioext"
 )
 
+func hasString(x string, y []string) bool {
+	for _, s := range y {
+		if s == x {
+			return true
+		}
+	}
+
+	return false
+}
+
 // Note: Constants testGroup, testCat, testTrue, testFalse, testPrintDir, and
 //       testSleep should be defined per platform
 
 func TestSystem(t *testing.T) {
+	var err error
+	var u *User
+
+	currentUser, err := user.Current()
+	assert.NoError(t, err)
+
+	if currentUser.Username != "root" {
+		t.Skip("This test requires root privileges, skipping.")
+	}
+
 	// Setup temporary home directory
 	homeDir := filepath.Join(os.TempDir(), slugid.Nice())
 	require.NoError(t, os.MkdirAll(homeDir, 0777))
 	defer os.RemoveAll(homeDir)
 
-	var u *User
-	var err error
-
 	t.Run("CreateUser", func(t *testing.T) {
-		u, err = CreateUser(homeDir, nil)
+		groups := []*Group{}
+		for _, groupName := range testGroups {
+			var g *user.Group
+			var gid int
+
+			g, err = user.LookupGroup(groupName)
+			require.NoError(t, err)
+
+			gid, err = strconv.Atoi(g.Gid)
+			require.NoError(t, err)
+
+			groups = append(groups, &Group{gid})
+		}
+
+		u, err = CreateUser(homeDir, groups)
+		require.NoError(t, err)
+		su, err := user.Lookup(u.name)
+		require.NoError(t, err)
+
+		// some sanity checks
+		require.EqualValues(t, u.name, su.Username)
+
+		uid, err := strconv.ParseUint(su.Uid, 10, 32)
+		require.NoError(t, err)
+		require.EqualValues(t, u.uid, uint32(uid))
```

Comment:
platform specific and accessing private variables...

If we want to do whitebox tests, let's make a `system_darwin_test.go` file and add darwin specific test cases in there...
Ideally, we can have both generic blackbox tests and platform specific whitebox tests. IMO, the whitebox testing is probably overkill, but it doesn't hurt to have it :)